Added system-specific bindings for netinet/in.h

jmdavis opened the pull request:
Both FreeBSD and Linux add quite a few items to netinet/in.h which are not part of POSIX (some being part of other standards, some not). I'd be shocked if the other *nix systems out there didn't do the same, but I don't have any such systems set up, so I haven't dealt with them. This adds core.sys.{freebsd,linux}.netinet.in_ which include the non-POSIX declarations and publicly import core.sys.posix.netinet.in_. It includes quite a few of the constants that are in std.c.linux.socket (which is deprecated and overdue for removal), including ones that vibe.d uses and has never bothered to redeclare itself or add to druntime in order to not use the deprecated symbols. With this, hopefully, it can stop using the deprecated symbols.

Not all of the symbols from those header files have been included (some were marked as deprecated or obsolete and some have been included as commented out, because they required symbols that we don't have yet or were macros which were too much of a pain to figure out how to port at the moment), but most of the symbols have been included.

Also, I added a number of files to mak/SRCS that were missing for some reason.

nemanja-boric-sociomantic commented on mak/SRCS:
```diff
@@ -52,8 +52,19 @@ SRCS=\
 	src\core\sync\rwmutex.d \
```
Re: commit message: Did you forget to include change to `pthread_np` into the patch?

nemanja-boric-sociomantic replied:
(and here)

jmdavis replied:
No. I removed it purposefully, because I was afraid that it would break something to just change the module name even though it was wrong. I just forgot to update the commit message. So, I'll fix that shortly. I'll figure out what to do about pthread_np in a different PR.

nemanja-boric-sociomantic commented on src/core/sys/freebsd/netinet/in_.d:
```diff
@@ -0,0 +1,581 @@
+//Written in the D programming language
```
Is this line really necessary, feels like noise :-) ?

jmdavis replied:
We normally put that at the top of all of the D files. AFAIK, all of the .d files in druntime and Phobos have it.

nemanja-boric-sociomantic replied:
Hmm, looks like it's half/half for phobos, and 23 vs 450 for druntime (that's why it looked strange to me), but fair enough :+1:

jmdavis replied:
I think that Walter likes people to put in on their files for searchability when google comes trauling along...

wilzbach replied:
Yes, Walter insists on this whenever he happens to catch a PR with a new module addition.
I don't think it makes any difference in Google optimization nowadays, but it definitely doesn't hurt.

nemanja-boric-sociomantic commented on src/core/sys/freebsd/netinet/in_.d:
```diff
@@ -0,0 +1,581 @@
+        enum IPV6PORT_RESERVEDMIN = 600;
+        enum IPV6PORT_RESERVEDMAX = IPV6PORT_RESERVED - 1;
+
+// TODO
```
TODO on purpose?

jmdavis replied:
I guess that I can add a note about it if that helps, but yes, on purpose. I didn't want to take the time to figure out how to convert it, though thinking about it now, I'm pretty sure that I know how. So, I'll see if I'm right and update it if so and comment it for later if not. The one on the Linux side is for the same reason

nemanja-boric-sociomantic replied:
I was just curious if you left this for latter and then forgot (which happens to me on a regular basis). No, I don't think there's need to add this note - after the review is finished, I don't think the note would carry any information.

jmdavis replied:
I just added the implementation. It was pretty straightforward. For some reason, I had been thinking that something more involved was going to be required.

nemanja-boric-sociomantic commented on src/core/sys/linux/netinet/in_.d:
```diff
@@ -0,0 +1,387 @@
+enum INADDR_ALLRTRS_GROUP   = 0xe0000002;
+enum INADDR_MAX_LOCAL_GROUP = 0xe00000ff;
+
+// TODO
```
ditto todo

jmdavis commented:
Okay. I managed to get the OS X headers from someone so that I was able to add them into the mix, though I obviously couldn't run the OS X build locally, so I may have to deal with some autotester failures before I get it compiling on OS X. I didn't realize that the OS X stuff had been moved to core.sys.darwin, though the dmd PR where that was discussed makes all of that sound like a mess. :|  In any case, I just made them work for `version(OSX)` for now, since while I expect that they're valid for the other "darwin" OSes, I don't know for sure. Someone who actually deals with those OSes will have to take care of that.

I also changed it so that linux uses `version(CRuntime_Glibc)` now, since it seems that that's what we used for core.sys.posix.netinet.in_, though based on what I understand of the reasoning for using `linux` verses `CRuntime_Glibc`, it seems like it almost makes `linux` useless, because applications almost always deal with libc and not the kernel. :|

jacob-carlborg commented:
@jmdavis Do you know which version of the SDK the macOS headers are from?

jacob-carlborg commented:
I diffed the macOS headers and compared to all the other Apple platforms in the Xcode 9 SDK (still in beta) . There are no differences.

jacob-carlborg commented on src/core/sys/darwin/netinet/in_.d:
```diff
@@ -0,0 +1,605 @@
+    // and on OS X. So, while it _might_ be a bug, if it is, it's an upstream
+    // one, and we're compatible with it.
+    extern(D) bool IN_EXPERIMENTAL(in_addr_t i) pure @safe { return (i & 0xf0000000) == 0xf0000000; }
+    extern(D) bool IN_BADCLASS(in_addr_t i) pure @safe { return (i & 0xf0000000) == 0xf0000000; }
```
These are equal in the header file as well.

jmdavis replied:
That's what the comment says.

jacob-carlborg commented on src/core/sys/darwin/netinet/in_.d:
```diff
@@ -0,0 +1,605 @@
+    // one, and we're compatible with it.
+    extern(D) bool IN_EXPERIMENTAL(in_addr_t i) pure @safe { return (i & 0xf0000000) == 0xf0000000; }
+    extern(D) bool IN_BADCLASS(in_addr_t i) pure @safe { return (i & 0xf0000000) == 0xf0000000; }
+
```
The following to constants are missing:

```c
#define	INADDR_LOOPBACK		(u_int32_t)0x7f000001

#define	INADDR_NONE		0xffffffff		/* -1 return */
```

jmdavis replied:
Because they're in `core.sys.posix.netinet._in`, because they're POSIX, unlike the stuff in this PR.

jacob-carlborg commented on src/core/sys/darwin/netinet/in_.d:
```diff
@@ -0,0 +1,605 @@
+
+    enum IN_LOOPBACKNET = 127;
+
+
```
This is missing:

```c
#define INET_ADDRSTRLEN                 16
```

jacob-carlborg commented on src/core/sys/darwin/netinet/in_.d:
```diff
@@ -0,0 +1,605 @@
+    enum IP_PORTRANGE       = 19;
+    enum IP_RECVIF          = 20;
+
+    enum IP_IPSEC_POLICY = 21;
```
Missing: `#define	IP_FAITH		22   /* deprecated */`.

jmdavis replied:
Yes, because it says it's deprecated. I saw no point in including declarations where the header file said that they were deprecated or obsolete. A number of the declarations in in.h were marked that way, and some were marked that way on one system, so I left them out on all of the systems. I don't think that anything that's missing is not missing on purpose, and everything that you've mentioned as missing in a comment is purposefully missing. More declarations can be added later if we need to, whereas removing them later is more of a pain. I don't expect that we'll need the ones that I left out.

jacob-carlborg commented on src/core/sys/darwin/netinet/in_.d:
```diff
@@ -0,0 +1,605 @@
+        sockaddr_storage gsr_group;
+        sockaddr_storage gsr_source;
+    };
+
```
This is missing, but the header says it's private:

```c
/*
 * The following structure is private; do not use it from user applications.
 * It is used to communicate IP_MSFILTER/IPV6_MSFILTER information between
 * the RFC 3678 libc functions and the kernel.
 */
struct __msfilterreq {
	uint32_t		 msfr_ifindex;	/* interface index */
	uint32_t		 msfr_fmode;	/* filter mode for group */
	uint32_t		 msfr_nsrcs;	/* # of sources in msfr_srcs */
	uint32_t		__msfr_align;	
	struct sockaddr_storage	 msfr_group;	/* group address */
	struct sockaddr_storage	*msfr_srcs;
};
```

jacob-carlborg commented on src/core/sys/darwin/netinet/in_.d:
```diff
@@ -0,0 +1,605 @@
+// more sense to put them in here so that the way they're imported corresponds
+// with the correct way of #including them in C/C++.
+// =============================================================================
+
```
Missing, not sure if it's needed:

```c
/*
 * Identification of the network protocol stack
 * for *BSD-current/release: http://www.kame.net/dev/cvsweb.cgi/kame/COVERAGE
 * has the table of implementation/integration differences.
 */
#define	__KAME__
#define	__KAME_VERSION		"2009/apple-darwin"
```

jacob-carlborg commented on src/core/sys/darwin/netinet/in_.d:
```diff
@@ -0,0 +1,605 @@
+
+
+    enum IN6ADDR_ANY_INIT = in6_addr.init;
+    enum IN6ADDR_LOOPBACK_INIT = in6_addr([0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
```
Indentation is not correct.

jacob-carlborg commented on src/core/sys/darwin/netinet/in_.d:
```diff
@@ -0,0 +1,605 @@
+    enum IPV6PORT_RESERVEDMAX = IPV6PORT_RESERVED - 1;
+}
+
+
```
Missing:

```c
/*
 * Socket address for IPv6
 */
#if !defined(_POSIX_C_SOURCE) || defined(_DARWIN_C_SOURCE)
#define	SIN6_LEN
#endif /* (_POSIX_C_SOURCE && !_DARWIN_C_SOURCE) */
```

jacob-carlborg commented on src/core/sys/darwin/netinet/in_.d:
```diff
@@ -0,0 +1,605 @@
+{
+    enum IPV6_PORTRANGE    = 14;
+    enum ICMP6_FILTER      = 18;
+
```
Missing:

```c
#define	IPV6_2292PKTINFO	19 /* bool; send/recv if, src/dst addr */
#define	IPV6_2292HOPLIMIT	20 /* bool; hop limit */
#define	IPV6_2292NEXTHOP	21 /* bool; next hop addr */
#define	IPV6_2292HOPOPTS	22 /* bool; hop-by-hop option */
#define	IPV6_2292DSTOPTS	23 /* bool; destinaion option */
#define	IPV6_2292RTHDR		24 /* ip6_rthdr: routing header */

/* buf/cmsghdr; set/get IPv6 options [obsoleted by RFC3542] */
#define	IPV6_2292PKTOPTIONS	25

#ifdef __APPLE_USE_RFC_2292
#define	IPV6_PKTINFO	IPV6_2292PKTINFO
#define	IPV6_HOPLIMIT	IPV6_2292HOPLIMIT
#define	IPV6_NEXTHOP	IPV6_2292NEXTHOP
#define	IPV6_HOPOPTS	IPV6_2292HOPOPTS
#define	IPV6_DSTOPTS	IPV6_2292DSTOPTS
#define	IPV6_RTHDR	IPV6_2292RTHDR
#define	IPV6_PKTOPTIONS	IPV6_2292PKTOPTIONS
#endif /* __APPLE_USE_RFC_2292 */
```

jacob-carlborg commented on src/core/sys/darwin/netinet/in_.d:
```diff
@@ -0,0 +1,605 @@
+    enum IPV6_BINDV6ONLY = IPV6_V6ONLY;
+
+    enum IPV6_IPSEC_POLICY = 28;
+
```
Missing: `#define	IPV6_FAITH		29 /* deprecated */`.

jacob-carlborg commented on src/core/sys/darwin/netinet/in_.d:
```diff
@@ -0,0 +1,605 @@
+    enum IPV6_HOPOPTS  = 49;
+    enum IPV6_DSTOPTS  = 50;
+    enum IPV6_RTHDR    = 51;
+
```
Missing:

```c
#define	IPV6_PKTINFO	IPV6_3542PKTINFO
#define	IPV6_HOPLIMIT	IPV6_3542HOPLIMIT
#define	IPV6_NEXTHOP	IPV6_3542NEXTHOP
#define	IPV6_HOPOPTS	IPV6_3542HOPOPTS
#define	IPV6_DSTOPTS	IPV6_3542DSTOPTS
#define	IPV6_RTHDR	IPV6_3542RTHDR
```

jacob-carlborg commented on src/core/sys/darwin/netinet/in_.d:
```diff
@@ -0,0 +1,605 @@
+    enum IPV6_DONTFRAG = 62;
+
+    enum IPV6_PREFER_TEMPADDR = 63;
+
```
Missing:

```c
/*
 * The following option is private; do not use it from user applications.
 * It is deliberately defined to the same value as IP_MSFILTER.
 */
#define	IPV6_MSFILTER		74 /* struct __msfilterreq; */
#endif /* __APPLE_USE_RFC_3542 */
```

jacob-carlborg commented on src/core/sys/darwin/netinet/in_.d:
```diff
@@ -0,0 +1,605 @@
+    enum IPV6CTL_SOURCECHECK        = 10;
+    enum IPV6CTL_SOURCECHECK_LOGINT = 11;
+    enum IPV6CTL_ACCEPT_RTADV       = 12;
+
```
Missing: `#define	IPV6CTL_KEEPFAITH	13	/* deprecated */`.

jacob-carlborg commented on src/core/sys/darwin/netinet/in_.d:
```diff
@@ -0,0 +1,605 @@
+    enum ICMPV6CTL_ND6_ONLINKNSRFC4861 = 50;
+
+    enum IPV6CTL_MAXID = 51;
+
```
Missing:

```c
extern int inet6_option_space(int);
extern int inet6_option_init(void *, struct cmsghdr **, int);
extern int inet6_option_append(struct cmsghdr *, const __uint8_t *, int, int);
extern __uint8_t *inet6_option_alloc(struct cmsghdr *, int, int, int);
extern int inet6_option_next(const struct cmsghdr *, __uint8_t **);
extern int inet6_option_find(const struct cmsghdr *, __uint8_t **, int);
```

jmdavis replied:
Per the linux version of in.h, they come from an obsolete RFC, so I didn't put them in either the FreeBSD or OS X bindings.

jacob-carlborg commented:
I found it a bit difficult to review since some symbols that are defined in `in.h` is missing here but instead are defined in the Posix `in_.d` module.

jacob-carlborg commented on src/core/sys/darwin/netinet/in_.d:
```diff
@@ -0,0 +1,605 @@
+
+// These are probably the same on the other "Darwin" OSes, but someone will
+// need to verify that.
+version(OSX):
```
Verified, please change to `Darwin`.

jacob-carlborg commented on src/core/sys/darwin/netinet/in_.d:
```diff
@@ -0,0 +1,605 @@
+    enum IN_CLASSA_HOST   = 0x00ffffff;
+    enum IN_CLASSA_MAX    = 128;
+
+    extern(D) bool IN_CLASSB(in_addr_t i) pure @safe { return (i & 0xc0000000) == 0x80000000; }
```
All of these macros use `u_int32_t` instead of `in_addr_t `. They all correspond to the same type though.

jacob-carlborg replied:
Wouldn't a closer translation of these macros be?

```d
bool IN_CLASSB(T)(T i) { return (cast(uint) i & 0xc0000000) == 0x80000000; }
```

jmdavis replied:
I translated them the way that similar macros were translated in `core.sys.posix.netinet.in_`. Also, this way, they won't accidentally accept an `in6_addr`.

jmdavis commented on src/core/sys/darwin/netinet/in_.d:
```diff
@@ -0,0 +1,605 @@
+//Written in the D programming language
+
+/++
+    D header file for FreeBSD's extensions to POSIX's netinet/in.h.
```
Yay, for copy-paste errors... :|

jmdavis commented:
Okay. I fixed the typos in the OS X stuff, and switched it to use `version(Darwin)` as requested. As mentioned response to some of the comments, the missing declarations are missing on purpose - mostly because the declarations were marked as deprecated or obsolete, so in theory, no one should be using them, and it's easier to add more later than to remove them later.

jacob-carlborg commented:
In general I don't think it's our job to decide which symbols should or should not be included in bindings. We should just expose all the symbols. It's fine to remove internal or private symbols, but I do think we should include symbols for obsolete protocols, if they are present in the header file by default.

jacob-carlborg commented:
Wouldn't it be better to move all the symbols from the Posix file to the platform specific ones now when they exist? The Posix file is already versioned by platform and it's more complicated to maintain when some of the symbols are in one file and some in another file.

This would go in a separate pull request.

jacob-carlborg commented:
Since I requested changes, can I approve them now somewhere? Or can't I do that because I don't have write access?

nemanja-boric-sociomantic commented:
@jacob-carlborg you can approve them, go to "Files changed" tab and then "Review" in the top right corner.

jmdavis commented:
> In general I don't think it's our job to decide which symbols should or should not be included in bindings. We should just expose all the symbols. It's fine to remove internal or private symbols, but I do think we should include symbols for obsolete protocols, if they are present in the header file by default.

I don't agree, but you're free to create a PR to add the other symbols.

> Wouldn't it be better to move all the symbols from the Posix file to the platform specific ones now when they exist? The Posix file is already versioned by platform and it's more complicated to maintain when some of the symbols are in one file and some in another file.

And then how would someone import just the POSIX declarations? If code imports just the POSIX declarations and just uses those, it should be platform-independent between POSIX systems, whereas if it imports everything from a platform, then it risks accidentally using something that's not actually POSIX. Those who want or need to import the system-specific files obviously can, and this PR's whole purpose is to make that possible, but by having the POSIX stuff be separate, it makes it easier to avoid creating platform-dependent code by accident.

I wouldn't expect a lot of maintenance to go into these files anyway. Most of this stuff doesn't change, and it's likely that something will be added to these files only if a declaration is added to the C headers and someone wants to use them from D and finds that they aren't here. I very much doubt that anyone is going to be going to the effort of periodically trying to go over every declaration to make sure that the C and D declarations are in sync. And even if someone does want to do that, it's easy enough to grep for the symbol in question or just search `core.sys.posix.netinet.in_`. And the POSIX stuff is even less likely to change than the rest.

jacob-carlborg commented:
> And then how would someone import just the POSIX declarations?

That's a pretty good point. I didn't think of that.

> I wouldn't expect a lot of maintenance to go into these files anyway.

That's probably true. It just made my life a bit more difficult when I reviewed this 😃.

MartinNowak commented:
The added file is missing in mak/COPY.

> Also, I added a number of files to mak/SRCS that were missing for some reason.

Those haven't been there b/c they are header only files and don't need to be compiled :/.

> In general I don't think it's our job to decide which symbols should or should not be included in bindings. We should just expose all the symbols. It's fine to remove internal or private symbols, but I do think we should include symbols for obsolete protocols, if they are present in the header file by default.

Yes, headers should be one-to-one translations. In fact they should be diff-able against the C headers. You might want to reread the Deimos guidelines (https://jkm.github.io/phobos/deimos.html).

For OS-specific headers I've tried to establish to translate the full header, but comment out stuff that's Posix.
https://github.com/dlang/druntime/blob/77b70a76f474422d4375a6432944e51daca95edf/src/core/sys/linux/sys/mman.d
https://github.com/dlang/druntime/blob/77b70a76f474422d4375a6432944e51daca95edf/src/core/sys/linux/dlfcn.d

You might also have overlooked that the bits/ folder in glibc is for platform/asm specific parts, i.e. needs to be versioned for the translated platform/architecture. Never translate the headers from your host OS, instead use glibc's source and please also add links to their glibc.git.
